### PR TITLE
Connect transform dialect passes into IREE.

### PR DESCRIPTION
**Pull request by nicolasvasilache**
This revision provides a first connection between IREE and the transform dialect by
allowing the orthogonal specification of a linalg_transform file (policy) that is
parsed and applied on the fly.

### CODE REVIEW

**Comment by nicolasvasilache** [comment]
@benvanik could you please comment on at least the extra dependencies ?
I had to thread dependencies on PDL and PDLInterp though a few layers and I am unclear this is the best way.

OTOH, in order to create a more contained unit test for this PR, I also had to do the same for the Bufferization dialect.

The nested pass structure make the additional changes less than ideal but maybe there is a better / more automated way to achieve the same?

File: iree/compiler/Codegen/LLVMCPU/test/linalg_transform.mlir [hunk 1]
```diff
@@ -0,0 +1,32 @@
+// RUN: iree-opt %s  -pass-pipeline='hal.executable(hal.executable.variant(iree-llvmcpu-lower-executable-target))' -iree-codegen-use-linalg-transform-interp --iree-codegen-use-linalg-transform-interp --linalg-transform-file-name=%p/linalg_transform_spec.mlir | FileCheck %s
```
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/LLVMCPU/test/linalg_transform.mlir:1]: The flag is repeated twice. Probably not required?

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp [hunk 2]
```diff
@@ -146,8 +149,13 @@ class LLVMAOTTargetBackend final : public TargetBackend {
   void getDependentDialects(DialectRegistry &registry) const override {
     mlir::registerLLVMDialectTranslation(registry);
     // TODO: make inclusion of ArmNeon conditional?
-    registry
-        .insert<IREE::Codegen::IREECodegenDialect, arm_neon::ArmNeonDialect>();
+    // clang-format off
+    registry.insert<IREE::Codegen::IREECodegenDialect,
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp:153]: I am surprised you need this, cause adding these to `LLVMCPULowerExecutableTarget` should have been enough for using with `iree-opt`.

I see you already updated https://github.com/google/iree/blob/main/iree/tools/init_mlir_dialects.h. THat should have been enough. Do you know what fails if you dont add these here.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Overall looks fine. Just a few questions about the dependent dialect specification.

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp (hunk 2 above)
Comment (nicolasvasilache) [reply iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp:153]: The first test I had was using iree-translate / iree-compile.
I can remove here for the sake of being more self-contained but things will not work end to end for now.
I'll reactivate when I reconnect e2e execution.

File: iree/tools/init_iree_dialects.h [hunk 3]
```diff
@@ -37,6 +38,7 @@ inline void registerIreeDialects(DialectRegistry &registry) {
                   IREE::Flow::FlowDialect,
                   IREE::HAL::HALDialect,
                   IREE::LinalgExt::IREELinalgExtDialect,
+                  mlir::linalg::transform::LinalgTransformDialect,
```
Comment (MaheshRavishankar) [inline iree/tools/init_iree_dialects.h:41]: Ah, I know why this is here, but the namespace difference is a bit off-putting. Just noting, not making it blocking.

**Review by MaheshRavishankar — APPROVED** [review]

**Pull request merged**